src/misc/slaughter.ts

```diff
@@ -0,0 +1,219 @@
+
+const slaughterableAnimals = new Map<string, Array<LootDrop<undefined>>>();
+const slaughterTools: Material[] = [Material.IRON_SWORD, Material.IRON_AXE];
+const slaugtherSound = Sound.BLOCK_SLIME_BLOCK_STEP;
+
+addSlaughterableAnimal(EntityType.COW, [
+  { item: Material.LEATHER, rarity: 0.3, count: 1 },
+  { item: Material.BEEF, rarity: 0.5, count: 1 },
+  { item: Material.BONE, rarity: 0.2, count: 1 },
+]);
+
+addSlaughterableAnimal(EntityType.PIG, [
+  { item: Material.PORKCHOP, rarity: 0.5, count: 1 },
+  { item: Material.BONE, rarity: 0.2, count: 1 },
+]);
+
+addSlaughterableAnimal(EntityType.CHICKEN, [
+  { item: Material.CHICKEN, rarity: 0.6, count: 1 },
+  { item: Material.FEATHER, rarity: 0.8, count: 1 },
+]);
+
+/**
+ * Applies new animal to slaughterable map if it isn't already a key
+ * @param entityType Animal which will be slaughterable
+ * @param drops Animal drops
+ */
+function addSlaughterableAnimal(
+  entityType: EntityType,
+  drops: LootDrop<any>[],
+) {
+  if (slaughterableAnimals.has(entityType.toString())) return;
+
+  slaughterableAnimals.set(entityType.toString(), drops);
+}
+
+/**
+ * Returns list of animal drops based on entity type after each
+ * hit on animal corpse.
+ * @param entityType EntityType to determine drops
+ */
+function getAnimalDrops(entityType: EntityType): ItemStack[] {
+  let drops: ItemStack[] = [];
+  const dropsFromMap = slaughterableAnimals.get(entityType.toString());
+  if (dropsFromMap) drops = generateLoot(undefined, dropsFromMap);
+  return drops;
+}
+
+/**
+ * Turns LivingEntity into animal corpse
+ * @param entity LivingEntity which will be turned into animal corpse
+ */
+function createAnimalCorpse(entity: LivingEntity) {
+  entity.setAI(false);
+  entity.setCustomName(CORPSE_NAME);
+  entity.setSilent(true);
+  // Add armor stand as passenger to hide nametag
+  if (entity.passenger === null) {
+    const nameHider = entity.world.spawnEntity(
+      entity.location.add(0, 200, 0),
+      EntityType.ARMOR_STAND,
+    ) as ArmorStand;
+    nameHider.setVisible(false);
+    nameHider.setMarker(true);
+    nameHider.setInvulnerable(true);
+    entity.setPassenger(nameHider);
+  }
+  // Set animal health to match hit amount
+  const hitsToRemove = randomInt(
+    MIN_BODY_HITS <= entity.maxHealth ? MIN_BODY_HITS : entity.maxHealth,
+    MAX_BODY_HITS <= entity.maxHealth ? MAX_BODY_HITS : entity.maxHealth,
+  );
+  entity.setHealth(hitsToRemove);
+  dropBodyUntilOnGround(entity);
+}
+
+/**
+ * Drops animal loot after each hit
+ * @param entity Animal corpse
+ */
+function handleDrops(entity: LivingEntity) {
+  const drops = getAnimalDrops(entity.type);
+  for (const item of drops) {
+    entity.world.dropItemNaturally(entity.location, item);
+  }
+}
+
+/**
+ * Adds downwards velocity to entity until it hits the ground.
+ * Calls itself recursively.
+ * @param entity Entity to be dropped to the ground
+ */
+function dropBodyUntilOnGround(entity: LivingEntity) {
```

Comment:
Consider async function with loop with <code>await wait(200, 'millis')</code> instead of recursion with setTimeout. It is easier to read IMO.